docs/rfcs/20220413-aot-for-all-snode.md

```diff
@@ -0,0 +1,380 @@
+
+# Background
+
+Currently, Taichi fields are defined and used in the following manner:
+
+```py
+a = ti.field(ti.i32)
+b = ti.field(ti.f32)
+ti.root.pointer(ti.ij, 16).dense(ti.ij, 16).place(a, b)
+
+@ti.kernel
+def run():
+  for I in ti.grouped(a):
+    b[I] = a[I] * 4.2
+```
+
+While this is convenient for Python users, it imposes challenges for the deployment side.
+
+1. Taichi fields are currently implemented as global variables.
+
+    This would result in the Taichi kernels being "not pure" and relying on implicit information. When saving such kernels into the AOT module, it is also necessary to save all the depdendant global states. Ideally, users should be able to create Taichi fields, and pass them into Taichi kernels as parameters.
+
+2. SNodes types are missing from the AOT module.
+
+    To moving towards the direction of passing fields into kernels, field and SNode types need to be saved into the AOT module as well.
+
+3. Fields data are not managed by the users.
+
+    Because fields are global, the Taichi runtime have to create and manage them. By localizing the fields, decoupling them from Taichi kernels, users can manage the memory resources for these fields.
+
+# Goals
+
+* Provide a SNode API that allows SNodes and Taichi fields to be localized, so that Taichi kernels can be made *pure*.
+* Supports describing the complete SNode tree type explicitly.
+* Make SNode types serializable into AOT, so that AOT can use all kinds of SNodes.
+* The new SNode API should offer compatibility with the existing usage.
+* (Uncertain, but highly desired) Decouple the element type from the SNode type, i.e. the situation where matrix fields has to been implemented in the "scattered" way to support SoA layout.
+
+## Non-Goals
+
+* Expand the support for sparse SNodes beyond LLVM's codegen, especially SPIR-V.
+
+# Detailed Design
+
+## A first attempt
+
+This API would clearly allow us to pass in fields into kernels:
+
+```py
+a = ti.field(ti.i32)
+b = ti.field(ti.f32)
+ti.root.pointer(ti.ij, 16).dense(ti.ij, 16).place(a, b)
+
+@ti.kernel
+def run(a: ?, b: ?):
+  for I in ti.grouped(a):
+    b[I] = a[I] * 4.2
+
+run(a, b)
+```
+
+However, it doesn't really work for AOT, because `a` and `b` are **attributes of a tree type**. That is, you *cannot* dump `a` and `b`'s types separately.
+
+To understand this problem, we can define something equivalent in C++:
+
+```cpp
+struct AB {
+  int32_t a;
+  float b;
+};
+
+using TreeType = PointerDense<AB>;
+```
+
+We cannot declare the `run` kernel as `void run(? a, ? b)`. Instead, we have to pass in a `TreeType` instance as a whole into the kernel, i.e., `void run(TreeType &tree)`.
+
+Internally, as you are using Taichi's SNode system to construct hierarchies, you are also constructing a `SNodeTree` type at the same time. This is done by Taichi's [`FieldsBuilder`](https://github.com/taichi-dev/taichi/blob/master/python/taichi/_snode/fields_builder.py).
+
+## A working design
```

Comment:
Would be great if you could share some feedback :-) @bsavery

Reply:
So overall, it seems to make sense.  However, I find the proposed syntax.... a bit "un-pythonic" to be honest.  

What about something (similar to python `dataclasses`) ?

```
@ti.tree_builder
class MySphereTreeBuilder:
    pos: ti.vec3
    radius: ti.f32
```

where you define the fields as a data structure in the "tree builder" class type

Reply:
Nice suggesteion! I'm not exactly sure how field view is gonna be handled here, but I feel like this is definitely possible to do with the current design.